File: android-core/src/main/java/com/mparticle/internal/AppStateManager.java
package com.mparticle.internal;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Application;
import android.content.ComponentName;
import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.SystemClock;

import com.mparticle.MPEvent;
import com.mparticle.MParticle;
import com.mparticle.identity.IdentityApi;
import com.mparticle.identity.IdentityApiRequest;
import com.mparticle.identity.MParticleUser;
import com.mparticle.internal.listeners.InternalListenerManager;

import org.json.JSONObject;

import java.lang.ref.WeakReference;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;


/**
 * This class is responsible for maintaining the session state by listening to the Activity lifecycle.
 */
public class  AppStateManager {

    private ConfigManager mConfigManager;
    Context mContext;
    private final SharedPreferences mPreferences;
    private InternalSession mCurrentSession = new InternalSession();
    private WeakReference<Activity> mCurrentActivityReference = null;

    private String mCurrentActivityName;
    /**
     * This boolean is important in determining if the app is running due to the user opening the app,
     * or if we're running due to the reception of a Intent such as an FCM message.
     */
    public static boolean mInitialized;

    AtomicLong mLastStoppedTime;
    /**
     * it can take some time between when an activity stops and when a new one (or the same one on a configuration change/rotation)
     * starts again, so use this handler and ACTIVITY_DELAY to determine when we're *really" in the background
     */
    Handler delayedBackgroundCheckHandler = new Handler();
    static final long ACTIVITY_DELAY = 1000;


    /**
     * Some providers need to know for the given session, how many 'interruptions' there were - how many
     * times did the user leave and return prior to the session timing out.
     */
    AtomicInteger mInterruptionCount = new AtomicInteger(0);

    /**
     * Constants used by the messaging/push framework to describe the app state when various
     * interactions occur (receive/show/tap).
     */
    public static final String APP_STATE_FOREGROUND = "foreground";
    public static final String APP_STATE_BACKGROUND = "background";
    public static final String APP_STATE_NOTRUNNING = "not_running";

    /**
     * Important to determine foreground-time length for a given session.
     * Uses the system-uptime clock to avoid devices which wonky clocks, or clocks
     * that change while the app is running.
     */
    private long mLastForegroundTime;

    boolean mUnitTesting = false;
    private MessageManager mMessageManager;
    private Uri mLaunchUri;
    private String mLaunchAction;

    public AppStateManager(Context context, boolean unitTesting) {
        mUnitTesting = unitTesting;
        mContext = context.getApplicationContext();
        mLastStoppedTime = new AtomicLong(getTime());
        mPreferences = context.getSharedPreferences(Constants.PREFS_FILE, Context.MODE_PRIVATE);
        ConfigManager.addMpIdChangeListener(new IdentityApi.MpIdChangeListener() {
            @Override
            public void onMpIdChanged(long newMpid, long previousMpid) {
                if (mCurrentSession != null) {
                    mCurrentSession.addMpid(newMpid);
                }
            }
        });
    }

    public AppStateManager(Context context) {
        this(context, false);
    }

    public void init(int apiVersion) {
        if (apiVersion >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            setupLifecycleCallbacks();
        }
    }

    public String getLaunchAction() {
        return mLaunchAction;
    }

    public Uri getLaunchUri() {
        return mLaunchUri;
    }

    public void setConfigManager(ConfigManager manager) {
        mConfigManager = manager;
    }

    public void setMessageManager(MessageManager manager) {
        mMessageManager = manager;
    }

    private long getTime() {
        if (mUnitTesting) {
            return System.currentTimeMillis();
        } else {
            return SystemClock.elapsedRealtime();
        }
    }

    public void onActivityResumed(Activity activity) {
        try {
            mCurrentActivityName = AppStateManager.getActivityName(activity);

            int interruptions = mInterruptionCount.get();
            if (!mInitialized || !getSession().isActive()) {
                mInterruptionCount = new AtomicInteger(0);
            }
            String previousSessionPackage = null;
            String previousSessionUri = null;
            String previousSessionParameters = null;
            if (activity != null) {
                ComponentName callingApplication = activity.getCallingActivity();
                if (callingApplication != null) {
                    previousSessionPackage = callingApplication.getPackageName();
                }
                if(activity.getIntent() != null) {
                    previousSessionUri = activity.getIntent().getDataString();
                    if (mLaunchUri == null) {
                        mLaunchUri = activity.getIntent().getData();
                    }
                    if (mLaunchAction == null) {
                        mLaunchAction = activity.getIntent().getAction();
                    }
                    if (activity.getIntent().getExtras() != null && activity.getIntent().getExtras().getBundle(Constants.External.APPLINK_KEY) != null) {
                        JSONObject parameters = new JSONObject();
                        try {
                            parameters.put(Constants.External.APPLINK_KEY, MPUtility.wrapExtras(activity.getIntent().getExtras().getBundle(Constants.External.APPLINK_KEY)));
                        } catch (Exception e) {

                        }
                        previousSessionParameters = parameters.toString();
                    }
                }
            }

            mCurrentSession.updateBackgroundTime(mLastStoppedTime, getTime());

            boolean isBackToForeground = false;
            if (!mInitialized) {
                initialize(mCurrentActivityName, previousSessionUri, previousSessionParameters, previousSessionPackage);
            } else if (isBackgrounded() && mLastStoppedTime.get() > 0) {
                isBackToForeground = true;
                mMessageManager.postToMessageThread(new CheckAdIdRunnable());
                logStateTransition(Constants.StateTransitionType.STATE_TRANS_FORE,
                        mCurrentActivityName,
                        mLastStoppedTime.get() - mLastForegroundTime,
                        getTime() - mLastStoppedTime.get(),
                        previousSessionUri,
                        previousSessionParameters,
                        previousSessionPackage,
                        interruptions);
            }
            mLastForegroundTime = getTime();

            if (mCurrentActivityReference != null) {
                mCurrentActivityReference.clear();
                mCurrentActivityReference = null;
            }
            mCurrentActivityReference = new WeakReference<Activity>(activity);

            MParticle instance = MParticle.getInstance();
            if (instance != null) {
                if (instance.isAutoTrackingEnabled()) {
                    instance.logScreen(mCurrentActivityName);
                }
                if (isBackToForeground) {
                    instance.Internal().getKitManager().onApplicationForeground();
                    Logger.debug("App foregrounded.");
                }
                instance.Internal().getKitManager().onActivityResumed(activity);
            }
        } catch (Exception e) {
                Logger.verbose("Failed while trying to track activity resume: " + e.getMessage());
        }
    }

    public void onActivityPaused(Activity activity) {
        try {
            mPreferences.edit().putBoolean(Constants.PrefKeys.CRASHED_IN_FOREGROUND, false).apply();
            mLastStoppedTime = new AtomicLong(getTime());
            if (mCurrentActivityReference != null && activity == mCurrentActivityReference.get()) {
                mCurrentActivityReference.clear();
                mCurrentActivityReference = null;
            }

            delayedBackgroundCheckHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    try {
                        if (isBackgrounded()) {
                            checkSessionTimeout();
                            logBackgrounded();
                            mConfigManager.setPreviousAdId();
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }, ACTIVITY_DELAY);

            MParticle instance = MParticle.getInstance();
            if (instance != null) {
                if (instance.isAutoTrackingEnabled()) {
                    instance.logScreen(
                            new MPEvent.Builder(AppStateManager.getActivityName(activity))
                                    .internalNavigationDirection(false)
                                    .build()
                    );
                }
                instance.Internal().getKitManager().onActivityPaused(activity);
            }
        } catch (Exception e) {
            Logger.verbose("Failed while trying to track activity pause: " + e.getMessage());
        }
    }

    public void ensureActiveSession() {
        if (!mInitialized) {
            initialize(null, null, null, null);
        }
        InternalSession session = getSession();
        session.mLastEventTime = System.currentTimeMillis();
        if (!session.isActive()) {
            newSession();
        } else {
            mMessageManager.updateSessionEnd(getSession());
        }
    }

    void logStateTransition(String transitionType, String currentActivity, long previousForegroundTime, long suspendedTime, String dataString, String launchParameters, String launchPackage, int interruptions) {
        if (mConfigManager.isEnabled()) {
            ensureActiveSession();
            mMessageManager.logStateTransition(transitionType,
                    currentActivity,
                    dataString,
                    launchParameters,
                    launchPackage,
                    previousForegroundTime,
                    suspendedTime,
                    interruptions
            );
        }
    }

    public void logStateTransition(String transitionType, String currentActivity) {
        logStateTransition(transitionType, currentActivity, 0, 0, null, null, null, 0);
    }

    /**
     * Creates a new session and generates the start-session message.
     */
    private void newSession() {
        startSession();
        mMessageManager.startSession(mCurrentSession);
        Logger.debug("Started new session");
        mMessageManager.startUploadLoop();
        enableLocationTracking();
        checkSessionTimeout();
    }

    private void enableLocationTracking() {
        if (mPreferences.contains(Constants.PrefKeys.LOCATION_PROVIDER)) {
            String provider = mPreferences.getString(Constants.PrefKeys.LOCATION_PROVIDER, null);
            long minTime = mPreferences.getLong(Constants.PrefKeys.LOCATION_MINTIME, 0);
            long minDistance = mPreferences.getLong(Constants.PrefKeys.LOCATION_MINDISTANCE, 0);
            if (provider != null && minTime > 0 && minDistance > 0) {
                MParticle instance = MParticle.getInstance();
                if (instance != null) {
                    instance.enableLocationTracking(provider, minTime, minDistance);
                }
            }
        }
    }

    boolean shouldEndSession() {
        InternalSession session = getSession();
        MParticle instance = MParticle.getInstance();
        return 0 != session.mSessionStartTime &&
                isBackgrounded()
                && session.isTimedOut(mConfigManager.getSessionTimeout())
                && (instance == null || !instance.Media().getAudioPlaying());
    }

    private void checkSessionTimeout() {
        delayedBackgroundCheckHandler.postDelayed( new Runnable() {
            @Override
            public void run() {
                if (shouldEndSession()) {
                    Logger.debug("Session timed out");
                    endSession();
                }
            }
        }, mConfigManager.getSessionTimeout());
    }

    private void initialize(String currentActivityName, String previousSessionUri, String previousSessionParameters, String previousSessionPackage) {
        mInitialized = true;
        logStateTransition(Constants.StateTransitionType.STATE_TRANS_INIT,
                currentActivityName,
                0,
                0,
                previousSessionUri,
                previousSessionParameters,
                previousSessionPackage,
                0);
    }

    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onActivityCreated(activity, savedInstanceState);
        }
    }

    public void onActivityStarted(Activity activity) {
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onActivityStarted(activity);
        }
    }

    public void onActivityStopped(Activity activity) {
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onActivityStopped(activity);
        }
    }

    private void logBackgrounded() {
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            logStateTransition(Constants.StateTransitionType.STATE_TRANS_BG, mCurrentActivityName);
            instance.Internal().getKitManager().onApplicationBackground();
            mCurrentActivityName = null;
            Logger.debug("App backgrounded.");
            mInterruptionCount.incrementAndGet();
        }
    }

    @TargetApi(14)
    private void setupLifecycleCallbacks() {
        ((Application) mContext).registerActivityLifecycleCallbacks(new MPLifecycleCallbackDelegate(this));
    }

    public boolean isBackgrounded() {
        return !mInitialized || (mCurrentActivityReference == null && (getTime() - mLastStoppedTime.get() >= ACTIVITY_DELAY));
    }

    private static String getActivityName(Activity activity) {
        return activity.getClass().getCanonicalName();
    }

    public String getCurrentActivityName() {
        return mCurrentActivityName;
    }

    public InternalSession getSession() {
        return mCurrentSession;
    }

    public void endSession() {
        Logger.debug("Ended session");
        mMessageManager.endSession(mCurrentSession);
        disableLocationTracking();
        mCurrentSession = new InternalSession();
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onSessionEnd();
        }
        InternalListenerManager.getListener().onSessionUpdated(mCurrentSession);
    }

    private void disableLocationTracking() {
        SharedPreferences.Editor editor = mPreferences.edit();
        editor.remove(Constants.PrefKeys.LOCATION_PROVIDER)
                .remove(Constants.PrefKeys.LOCATION_MINTIME)
                .remove(Constants.PrefKeys.LOCATION_MINDISTANCE)
                .apply();
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.disableLocationTracking();
        }
    }

    public void startSession() {
        mCurrentSession = new InternalSession().start(mContext);
        mLastStoppedTime = new AtomicLong(getTime());
        enableLocationTracking();
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onSessionStart();
        }
    }

    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onActivitySaveInstanceState(activity, outState);
        }
    }

    public void onActivityDestroyed(Activity activity) {
        MParticle instance = MParticle.getInstance();
        if (instance != null) {
            instance.Internal().getKitManager().onActivityDestroyed(activity);
        }
    }

    public WeakReference<Activity> getCurrentActivity() {
        return mCurrentActivityReference;
    }

    class CheckAdIdRunnable implements Runnable {

        @Override
        public void run() {
            String previousGoogleAdId = mConfigManager.getPreviousAdId();
            MPUtility.AdIdInfo adIdInfo =  MPUtility.getAdIdInfo(mContext);
            String currentGoogleAdId = adIdInfo == null ? null : adIdInfo.id;
            if (currentGoogleAdId != null && !currentGoogleAdId.equals(previousGoogleAdId)) {
                MParticle instance = MParticle.getInstance();
                if (instance != null) {
                    MParticleUser user = instance.Identity().getCurrentUser();

                    Builder builder;
                    if (user != null) {
                        builder = new Builder(user);
                    } else {
                        builder = new Builder();
                    }
                    instance.Identity().modify(builder
                            .googleAdId(currentGoogleAdId, previousGoogleAdId)
                            .build());
                }
            }
        }
    }

    class Builder extends IdentityApiRequest.Builder {
        Builder(MParticleUser user) {
            super(user);
        }
        Builder() {
            super();
        }

        @Override
        protected IdentityApiRequest.Builder googleAdId(String newGoogleAdId, String oldGoogleAdId) {
            return super.googleAdId(newGoogleAdId, oldGoogleAdId);
        }
    }
}

File: android-core/src/main/kotlin/com/mparticle/internal/AppStateManager.kt
package com.mparticle.internal

import android.annotation.TargetApi
import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import com.mparticle.MPEvent
import com.mparticle.MParticle
import com.mparticle.identity.IdentityApi.SingleUserIdentificationCallback
import com.mparticle.identity.IdentityApiRequest
import com.mparticle.identity.MParticleUser
import com.mparticle.internal.listeners.InternalListenerManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * This class is responsible for maintaining the session state by listening to the Activity lifecycle.
 */
open class AppStateManager @JvmOverloads constructor(
    context: Context,
    unitTesting: Boolean = false
) {
    private var mConfigManager: ConfigManager? = null
    var mContext: Context
    private val mPreferences: SharedPreferences
    open var session: InternalSession = InternalSession()

    open var currentActivity: WeakReference<Activity?>? = null
        set

    var currentActivityName: String? = null
        private set
    var mLastStoppedTime: AtomicLong

    /**
     * it can take some time between when an activity stops and when a new one (or the same one on a configuration change/rotation)
     * starts again, so use this handler and ACTIVITY_DELAY to determine when we're *really" in the background
     */
    @JvmField
    var delayedBackgroundCheckHandler: Handler = Handler(Looper.getMainLooper())

    /**
     * Some providers need to know for the given session, how many 'interruptions' there were - how many
     * times did the user leave and return prior to the session timing out.
     */
    var mInterruptionCount: AtomicInteger = AtomicInteger(0)

    /**
     * Important to determine foreground-time length for a given session.
     * Uses the system-uptime clock to avoid devices which wonky clocks, or clocks
     * that change while the app is running.
     */
    private var mLastForegroundTime: Long = 0

    var mUnitTesting: Boolean = false
    private var mMessageManager: MessageManager? = null
    open var launchUri: Uri? = null
         set
    var launchAction: String? = null
        private set

    init {
        mUnitTesting = unitTesting
        mContext = context.applicationContext
        mLastStoppedTime = AtomicLong(time)
        mPreferences = context.getSharedPreferences(Constants.PREFS_FILE, Context.MODE_PRIVATE)
        ConfigManager.addMpIdChangeListener { newMpid, previousMpid ->
            if (session != null) {
                session.addMpid(newMpid)
            }
        }
    }

    fun init(apiVersion: Int) {
        if (apiVersion >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            setupLifecycleCallbacks()
        }
    }

    fun setConfigManager(manager: ConfigManager?) {
        mConfigManager = manager
    }

    fun setMessageManager(manager: MessageManager?) {
        mMessageManager = manager
    }

    private val time: Long
        get() = if (mUnitTesting) {
            System.currentTimeMillis()
        } else {
            SystemClock.elapsedRealtime()
        }

    fun onActivityResumed(activity: Activity?) {
        try {
            currentActivityName = getActivityName(activity)

            val interruptions = mInterruptionCount.get()
            if (!mInitialized || !session.isActive) {
                mInterruptionCount = AtomicInteger(0)
            }
            var previousSessionPackage: String? = null
            var previousSessionUri: String? = null
            var previousSessionParameters: String? = null
            if (activity != null) {
                val callingApplication = activity.callingActivity
                if (callingApplication != null) {
                    previousSessionPackage = callingApplication.packageName
                }
                if (activity.intent != null) {
                    previousSessionUri = activity.intent.dataString
                    if (launchUri == null) {
                        launchUri = activity.intent.data
                    }
                    if (launchAction == null) {
                        launchAction = activity.intent.action
                    }
                    if (activity.intent.extras?.getBundle(Constants.External.APPLINK_KEY) != null) {
                        val parameters = JSONObject()
                        try {
                            parameters.put(
                                Constants.External.APPLINK_KEY,
                                MPUtility.wrapExtras(
                                    activity.intent.extras?.getBundle(Constants.External.APPLINK_KEY)
                                )
                            )
                        } catch (e: Exception) {
                            Logger.error("Exception on onActivityResumed ")
                        }
                        previousSessionParameters = parameters.toString()
                    }
                }
            }

            session.updateBackgroundTime(mLastStoppedTime, time)

            var isBackToForeground = false
            if (!mInitialized) {
                initialize(
                    currentActivityName,
                    previousSessionUri,
                    previousSessionParameters,
                    previousSessionPackage
                )
            } else if (isBackgrounded() && mLastStoppedTime.get() > 0) {
                isBackToForeground = true
                mMessageManager?.postToMessageThread(CheckAdIdRunnable(mConfigManager))
                logStateTransition(
                    Constants.StateTransitionType.STATE_TRANS_FORE,
                    currentActivityName,
                    mLastStoppedTime.get() - mLastForegroundTime,
                    time - mLastStoppedTime.get(),
                    previousSessionUri,
                    previousSessionParameters,
                    previousSessionPackage,
                    interruptions
                )
            }
            CoroutineScope(Dispatchers.IO).launch {
                mConfigManager?.setPreviousAdId()
            }
            mLastForegroundTime = time

            if (currentActivity != null) {
                currentActivity?.clear()
                currentActivity = null
            }
            currentActivity = WeakReference(activity)

            val instance = MParticle.getInstance()
            if (instance != null) {
                if (instance.isAutoTrackingEnabled) {
                    currentActivityName?.let {
                        instance.logScreen(it)
                    }
                }
                if (isBackToForeground) {
                    instance.Internal().kitManager.onApplicationForeground()
                    Logger.debug("App foregrounded.")
                }
                instance.Internal().kitManager.onActivityResumed(activity)
            }
        } catch (e: Exception) {
            Logger.verbose("Failed while trying to track activity resume: " + e.message)
        }
    }

    fun onActivityPaused(activity: Activity) {
        try {
            mPreferences.edit().putBoolean(Constants.PrefKeys.CRASHED_IN_FOREGROUND, false).apply()
            mLastStoppedTime = AtomicLong(time)
            if (currentActivity != null && activity === currentActivity?.get()) {
                currentActivity?.clear()
                currentActivity = null
            }

            delayedBackgroundCheckHandler.postDelayed(
                {
                    try {
                        if (isBackgrounded()) {
                            checkSessionTimeout()
                            logBackgrounded()
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                },
                ACTIVITY_DELAY
            )

            val instance = MParticle.getInstance()
            if (instance != null) {
                if (instance.isAutoTrackingEnabled) {
                    instance.logScreen(
                        MPEvent.Builder(getActivityName(activity))
                            .internalNavigationDirection(false)
                            .build()
                    )
                }
                instance.Internal().kitManager.onActivityPaused(activity)
            }
        } catch (e: Exception) {
            Logger.verbose("Failed while trying to track activity pause: " + e.message)
        }
    }

    fun ensureActiveSession() {
        if (!mInitialized) {
            initialize(null, null, null, null)
        }
        session.mLastEventTime = System.currentTimeMillis()
        if (!session.isActive) {
            newSession()
        } else {
            mMessageManager?.updateSessionEnd(this.session)
        }
    }

    fun logStateTransition(
        transitionType: String?,
        currentActivity: String?,
        previousForegroundTime: Long,
        suspendedTime: Long,
        dataString: String?,
        launchParameters: String?,
        launchPackage: String?,
        interruptions: Int
    ) {
        if (mConfigManager?.isEnabled == true) {
            ensureActiveSession()
            mMessageManager?.logStateTransition(
                transitionType,
                currentActivity,
                dataString,
                launchParameters,
                launchPackage,
                previousForegroundTime,
                suspendedTime,
                interruptions
            )
        }
    }

    fun logStateTransition(transitionType: String?, currentActivity: String?) {
        logStateTransition(transitionType, currentActivity, 0, 0, null, null, null, 0)
    }

    /**
     * Creates a new session and generates the start-session message.
     */
    private fun newSession() {
        startSession()
        mMessageManager?.startSession(session)
        Logger.debug("Started new session")
        mMessageManager?.startUploadLoop()
        enableLocationTracking()
        checkSessionTimeout()
    }

    private fun enableLocationTracking() {
        if (mPreferences.contains(Constants.PrefKeys.LOCATION_PROVIDER)) {
            val provider = mPreferences.getString(Constants.PrefKeys.LOCATION_PROVIDER, null)
            val minTime = mPreferences.getLong(Constants.PrefKeys.LOCATION_MINTIME, 0)
            val minDistance = mPreferences.getLong(Constants.PrefKeys.LOCATION_MINDISTANCE, 0)
            if (provider != null && minTime > 0 && minDistance > 0) {
                val instance = MParticle.getInstance()
                instance?.enableLocationTracking(provider, minTime, minDistance)
            }
        }
    }

    fun shouldEndSession(): Boolean {
        val instance = MParticle.getInstance()
        return (
                0L != session?.mSessionStartTime &&
                        isBackgrounded() &&
                        mConfigManager?.sessionTimeout?.let { session.isTimedOut(it) } == true &&
                        (instance == null || !instance.Media().audioPlaying)
                )
    }

    private fun checkSessionTimeout() {
        mConfigManager?.sessionTimeout?.toLong()?.let {
            delayedBackgroundCheckHandler.postDelayed({
                if (shouldEndSession()) {
                    Logger.debug("Session timed out")
                    endSession()
                }
            }, it)
        }
    }

    private fun initialize(
        currentActivityName: String?,
        previousSessionUri: String?,
        previousSessionParameters: String?,
        previousSessionPackage: String?
    ) {
        mInitialized = true
        logStateTransition(
            Constants.StateTransitionType.STATE_TRANS_INIT,
            currentActivityName,
            0,
            0,
            previousSessionUri,
            previousSessionParameters,
            previousSessionPackage,
            0
        )
    }

    fun onActivityCreated(activity: Activity?, savedInstanceState: Bundle?) {
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onActivityCreated(activity, savedInstanceState)
    }

    fun onActivityStarted(activity: Activity?) {
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onActivityStarted(activity)
    }

    fun onActivityStopped(activity: Activity?) {
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onActivityStopped(activity)
    }

    private fun logBackgrounded() {
        val instance = MParticle.getInstance()
        if (instance != null) {
            logStateTransition(Constants.StateTransitionType.STATE_TRANS_BG, currentActivityName)
            instance.Internal().kitManager.onApplicationBackground()
            currentActivityName = null
            Logger.debug("App backgrounded.")
            mInterruptionCount.incrementAndGet()
        }
    }

    @TargetApi(14)
    private fun setupLifecycleCallbacks() {
        (mContext as Application).registerActivityLifecycleCallbacks(
            MPLifecycleCallbackDelegate(
                this
            )
        )
    }

    open fun isBackgrounded(): Boolean {
        return !mInitialized || (currentActivity == null && (time - mLastStoppedTime.get() >= ACTIVITY_DELAY))
    }

    open fun fetchSession(): InternalSession {
        return session
    }

    fun endSession() {
        Logger.debug("Ended session")
        mMessageManager?.endSession(session)
        disableLocationTracking()
        session = InternalSession()
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onSessionEnd()
        InternalListenerManager.listener.onSessionUpdated(session)
    }

    private fun disableLocationTracking() {
        val editor = mPreferences.edit()
        editor.remove(Constants.PrefKeys.LOCATION_PROVIDER)
            .remove(Constants.PrefKeys.LOCATION_MINTIME)
            .remove(Constants.PrefKeys.LOCATION_MINDISTANCE)
            .apply()
        val instance = MParticle.getInstance()
        instance?.disableLocationTracking()
    }

    fun startSession() {
        session = InternalSession().start(mContext)
        mLastStoppedTime = AtomicLong(time)
        enableLocationTracking()
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onSessionStart()
    }

    fun onActivitySaveInstanceState(activity: Activity?, outState: Bundle?) {
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onActivitySaveInstanceState(activity, outState)
    }

    fun onActivityDestroyed(activity: Activity?) {
        val instance = MParticle.getInstance()
        instance?.Internal()?.kitManager?.onActivityDestroyed(activity)
    }

    internal class CheckAdIdRunnable(var configManager: ConfigManager?) : Runnable {
        override fun run() {
            val adIdInfo =
                MPUtility.getAdIdInfo(
                    MParticle.getInstance()?.Internal()?.appStateManager?.mContext
                )
            val currentAdId =
                (if (adIdInfo == null) null else (if (adIdInfo.isLimitAdTrackingEnabled) null else adIdInfo.id))
            val previousAdId = configManager?.previousAdId
            if (currentAdId != null && currentAdId != previousAdId) {
                val instance = MParticle.getInstance()
                if (instance != null) {
                    val user = instance.Identity().currentUser
                    if (user != null) {
                        instance.Identity().modify(
                            Builder(user)
                                .googleAdId(currentAdId, previousAdId)
                                .build()
                        )
                    } else {
                        instance.Identity()
                            .addIdentityStateListener(object : SingleUserIdentificationCallback() {
                                override fun onUserFound(user: MParticleUser) {
                                    instance.Identity().modify(
                                        Builder(user)
                                            .googleAdId(currentAdId, previousAdId)
                                            .build()
                                    )
                                }
                            })
                    }
                }
            }
        }
    }

    internal class Builder : IdentityApiRequest.Builder {
        constructor(user: MParticleUser?) : super(user)

        constructor() : super()

        public override fun googleAdId(
            newGoogleAdId: String?,
            oldGoogleAdId: String?
        ): IdentityApiRequest.Builder {
            return super.googleAdId(newGoogleAdId, oldGoogleAdId)
        }
    }

    companion object {
        /**
         * This boolean is important in determining if the app is running due to the user opening the app,
         * or if we're running due to the reception of a Intent such as an FCM message.
         */
        @JvmField
        var mInitialized: Boolean = false

        const val ACTIVITY_DELAY: Long = 1000

        /**
         * Constants used by the messaging/push framework to describe the app state when various
         * interactions occur (receive/show/tap).
         */
        const val APP_STATE_FOREGROUND: String = "foreground"
        const val APP_STATE_BACKGROUND: String = "background"
        const val APP_STATE_NOTRUNNING: String = "not_running"

        private fun getActivityName(activity: Activity?): String {
            return activity?.javaClass?.canonicalName ?: ""
        }
    }
}